refacto: migrate Crs to typescript

Desplandis opened the pull request:
## Description

This PR includes:
- Migration of `Crs` from javascript to typescript
- Fix of type issues (mainly undefined case not treated)
- Add a `ProjectionLike` type alias for named CRS which could be later extended as either a string or a `proj4` definition object (à la [OpenLayer](https://openlayers.org/en/latest/apidoc/module-ol_proj.html#~ProjectionLike))
- TSDoc compliant documentation
- Add more unit tests

## Motivation and Context

As described in proposal #2396, we aim to gradually migrate our entire codebase (with the exception of deprecated modules) from javascript to typescript. We choose to start with modules with no-dependency and move up in the dependency tree.

This PR is the first step of **Migrate geographic modules**, migrating the base module `Crs`. Follow-up steps includes migration of dependent modules `Coordinates/Ellipsoid` and `Extent`.

Desplandis committed:
refacto: migrate Crs to typescript

mgermerie commented on src/Core/Geographic/Crs.ts:
```diff
@@ -80,100 +85,92 @@ function toUnitWithError(crs) {
+     * @param crs - The CRS to validate.
      *
-     * @throws {Error} if the CRS is not valid.
+     * @throws {@link Error} if the CRS is not valid.
```
Are you sure about the `@link` ? Will it link to [this doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error) should `throws` be shown on doc template ?

Desplandis replied:
In `tsdoc` the [`@throws`](https://tsdoc.org/pages/tags/throws/) tag is purely informational so only contain text. It seems this usually start with an `{@link X}` block. We could use [`typedoc-plugin-mdn-links`](https://github.com/Gerrit0/typedoc-plugin-mdn-links`) to resolve those links.

mgermerie commented on src/Core/Geographic/Crs.ts:
```diff
@@ -80,100 +85,92 @@ function toUnitWithError(crs) {
      */
-    isValid(crs) {
+    isValid(crs: ProjectionLike) {
         toUnitWithError(crs);
```
This throws `No unit found for crs: '${crs}'` when this function should instead tell the CRS is not valid.

Desplandis replied:
Agree, we should maybe add a function that throws if a call to `proj4.defs` returns `undefined`.

mgermerie commented on src/Core/Geographic/Crs.ts:
```diff
@@ -80,100 +85,92 @@ function toUnitWithError(crs) {
      */
-    isGeographic(crs) {
+    isGeographic(crs: ProjectionLike) {
         return (toUnitWithError(crs) == UNIT.DEGREE);
```
Same here, the message thrown by `toUnitWithError` is not relevant.

mgermerie commented on src/Core/Geographic/Crs.ts:
```diff
@@ -80,100 +85,92 @@ function toUnitWithError(crs) {
-     * @return {boolean}
+     * @param crs - The CRS to test.
      */
     isGeocentric,
```
Not related to pure TS migration but on a code cleanliness consideration, we could define this the same way we export `isGeographic` or `reasonnableEpsilon` : we could either define them all outside of the `export DEFAULT` object and reference them here, or define them all directly in the `export DEFAULT` object.

Desplandis replied:
Better, in my opinion, we should just not use `export default` anymore in the codebase. There are not compatible with tree-shaking. Imagine we build a bundle importing `Crs.js` and just using the `isGeocentric`, with `export default` the whole code of `Crs.js` is bundled, with multiple `export` only this function is bundled.

Desplandis replied:
However, this may be outside the scope of this PR?

jailln replied:
I would be down for doing it in this PR and in general taking the opportunity of this ts migration to remove default exports.

jailln replied:
thanks :)

mgermerie commented on test/unit/crs.js:
```diff
@@ -34,13 +34,21 @@ describe('CRS assertions', function () {
     });
 
+    it('should assert that the CRS is geocentric', function () {
+        assert.ok(!CRS.isGeocentric('EPSG:4326'));
+        assert.ok(!CRS.isGeocentric('EPSG:7133'));
+        assert.ok(CRS.isGeocentric('EPSG:4978'));
+        assert.ok(!CRS.isGeocentric('EPSG:3857'));
```
For unit tests, is it really necessary to test three non geocentric CRS ?

mgermerie replied:
(And the same goes for all other tests, there is no benefits in testing the same result multiple times)

Desplandis replied:
Don't know, I just tried to have a test symmetric to the `isGeometric` one. I think we should test all four since two here are defined by the test (i.e. the user), one defined by `iTowns` in `Crs` and one by default by `proj4`.

mgermerie commented on src/Core/Geographic/Crs.ts:
```diff
@@ -55,22 +60,22 @@ function _unitFromProj4Unit(projunit) {
     mustBeString(crs);
     switch (crs) {
         case 'EPSG:4326' : return UNIT.DEGREE;
         case 'EPSG:4978' : return UNIT.METER;
```
We could delete these lines as the EPSG are defined (either in proj4 or in this file) and we could directly call the generic method as in `default` case to not create a specificity.

Desplandis replied:
Yeah agree.

mgermerie commented on src/Core/Geographic/Crs.ts:
```diff
@@ -55,22 +60,22 @@ function _unitFromProj4Unit(projunit) {
 }
 
-function toUnitWithError(crs) {
+function toUnitWithError(crs: ProjectionLike) {
```
Maybe we could rename it to a more explicit name, like `checkUnit` ? I admit I don't understand the basic role of the method without reading its code.

Desplandis replied:
The method instead of returning `unit | undefined` returns `unit` and throws if `undefined`. In my opinion, we should add a method to check whether a `proj4.defs` is defined.

Desplandis committed:
fix(babel): include ts files in prerequisites

Desplandis committed:
refacto(Crs): cleanup unit handling

jailln commented on package.json:
```diff
@@ -21,13 +21,13 @@
     "prepublishOnly": "npm run build && npm run transpile",
     "prepare": "cross-env NO_UPDATE_NOTIFIER=true node ./config/prepare.mjs && node ./config/replace.config.mjs",
-    "watch": "cross-env BABEL_DISABLE_CACHE=1 babel --watch src --out-dir lib",
+    "watch": "npm run transpile -- --watch",
```
one `--` too many?

Desplandis replied:
The added `--` is for avoiding to pass the `--watch` option to the `npm` command and instead pass it to the `babel` command in the `transpile` script.

jailln commented on src/Core/Geographic/Crs.ts:
```diff
@@ -0,0 +1,185 @@
+
+import type { ProjectionDefinition } from 'proj4';
+
+proj4.defs('EPSG:4978', '+proj=geocent +datum=WGS84 +units=m +no_defs');
```
Note that it is also defined in `Coordinates`. Can we remove it in `Coordinates`?

Desplandis replied:
I thought I removed it in Coordinates... Will do it!

P.S.: It was done in the subsequent PR https://github.com/iTowns/itowns/pull/2437

jailln commented on src/Core/Geographic/Crs.ts:
```diff
@@ -0,0 +1,185 @@
+/**
+ * A projection as a CRS identifier string.
+ */
+export type ProjectionLike = string;
```
Should we export it in the doc ? Also, the comment is not complete as it can either be a crs identifier or a wkt string describing the projection and passed directly to proj4 (in isGeocentric for instance)

Desplandis replied:
Yeah, we should export it in the documentation since this is the type of the argument of some functions of the `CRS` module.

I think the documentation is complete, this is an identifier referencing an already defined projection (I should maybe add this to the documentation). In `isGeocentric`, the call to `proj4.defs` is only for retrieving an already defined projection as an object of type `proj4.ProjectionDefinition`.

Desplandis replied:
Moreover, I added such type alias since we'll surely want in a near future to have a `ProjectionLike` either as an identifier or an `proj4.ProjectionDefinition`.

jailln replied:
Oh ok thanks, `proj4.defs` is quite misleading :sweat_smile: 

I just checked and it is not exported in the doc though, you may need to add a @typedef tag like for CameraTransformOptions? : `@typedef {Object} CameraTransformOptions`

Also, I'm wandering if it should appear in the Crs doc page or have its own page or be in a `Types` page, WDYT?

Desplandis replied:
The equivalent of typedef in typescript or either an `interface` or a `type`. Instead of being in the documentation, we can write the types directly.

jailln replied:
Yes but except if I'm missing something it is still not exported in the doc (see http://localhost:8080/docs/out/#api/Geographic/CRS after running `npm run doc`).

jailln replied:
Just saw your other comment :)

jailln commented on src/Core/Geographic/Crs.ts:
```diff
@@ -0,0 +1,185 @@
+    return !projection ? false : projection.projName == 'geocent';
+}
+
+function _unitFromProj4Unit(proj: ProjectionDefinition) {
```
I think we should remove the `_` since it is not exported anyway

Desplandis replied:
Removed.

jailln commented on src/Core/Geographic/Crs.ts:
```diff
@@ -0,0 +1,185 @@
+export default {
+    /**
+     * Units that can be used for a CRS.
+     */
```
Could you add possible values in the documentation please?

Desplandis replied:
Done.

jailln commented on src/Core/Geographic/Crs.ts:
```diff
@@ -0,0 +1,185 @@
+     * @param crs - The CRS to get the unit from.
+     * @returns Either `UNIT.METER`, `UNIT.DEGREE` or `undefined`.
+     */
+    toUnit,
```
I'd rather have all the exported method defined here for readability, wdyt?

Desplandis replied:
I named-exported all functions instead of using the default export.

jailln commented on src/Core/Geographic/Crs.ts:
```diff
@@ -0,0 +1,185 @@
+    isGeocentric,
+
+    /**
+     * Give a reasonnable epsilon to use with this CRS.
```
reasonable with one n

Desplandis replied:
Agree. Fixed it.

jailln reviewed (commented):
Thanks :) Nice to see that we are moving forward on this migration :+1: 

While we're at it, I would be down for renaming some functions to make their name clearer:
* `toUnit` -> `getUnit`
* `reasonnableEpsilon` -> `reasonableEpsilon`
* `formatToEPSG` -> `formatToEpsg` (to be consistent with capital letters in other methods). Or use capitals in all methods with EPSG and TMS.

I would also make private `is4326`.

Also, we could add a `isDegreeUnit` since we expose a `isMetricUnit`.

Desplandis committed:
refacto(Crs): use named exports instead of default export

Desplandis committed:
chore(Crs): update and refine documentation

Desplandis committed:
fix(Crs): correctly renamed reasonableEpsilon function

BREAKING CHANGE: CRS.reasonnableEspsilon renamed to CRS.reasonableEpsilon

Desplandis committed:
refacto(Crs): rename toUnit to getUnit

BREAKING CHANGE: CRS.toUnit renamed to CRS.getUnit

Desplandis commented:
@jailln I fixed all issues listed in your review.

I added an [`@internal`](https://tsdoc.org/pages/tags/internal/) tag for all functions that I deemed to be only used internally by itowns. This mainly includes all related epsg and tms functions, that may be removed soon since we split Extent and Tile. I only kept the TMS naming scheme for Tile to identify where Tile was used but we could use the EPSG naming scheme (removing the need for the format and test functions).

>     * `toUnit` -> `getUnit`
Done, shall I mark it as breaking change though?

>     * `formatToEPSG` -> `formatToEpsg` (to be consistent with capital letters in other methods). Or use capitals in all methods with EPSG and TMS.

I chose to kept the old name since we'll surely remove it due to the reasons above.
 
> Also, we could add a `isDegreeUnit` since we expose a `isMetricUnit`.
I think we could add it later if the need arise.

jailln commented:
Thanks for the changes :)

> @jailln I fixed all issues listed in your review.
> 
> I added an [`@internal`](https://tsdoc.org/pages/tags/internal/) tag for all functions that I deemed to be only used internally by itowns. This mainly includes all related epsg and tms functions, that may be removed soon since we split Extent and Tile. I only kept the TMS naming scheme for Tile to identify where Tile was used but we could use the EPSG naming scheme (removing the need for the format and test functions).

I would be clearer to only use the EPSG naming scheme yes.

> > ```
> > * `toUnit` -> `getUnit`
> > ```
> 
> Done, shall I mark it as breaking change though?

I'm not sure it is used but we might add a breaking change to the commit  message just in case.

Desplandis committed:
refacto(Crs): remove tms/epsg family of functions

BREAKING CHANGE: CRS.isEPSG and CRS.isTMS have been removed
BREAKING CHANGE: CRS.formatToESPG and CRS.formatToTMS have been removed

Desplandis commented:
Okay, I went a little bit further and removed the two naming schemes (which were not needed since #2412). This means that all `espg/tms` functions are not needed anymore, so I removed them too. However, shall I keep the `espg/tms` functions for backward compatibility (and deprecated them) but remove reference in the codebase?

I don't think they are used outside of the codebase IMO.

As for the documentation, `ProjectionLike` documentation is not generated with JSDoc but is generated with TypeDoc. I however did change our documentation engine for now, since it will introduce a lot of conflicts with #2414. Waiting for it to be merged!

jailln commented:
> Okay, I went a little bit further and removed the two naming schemes (which were not needed since #2412). This means that all `espg/tms` functions are not needed anymore, so I removed them too. However, shall I keep the `espg/tms` functions for backward compatibility (and deprecated them) but remove reference in the codebase?
> 
> I don't think they are used outside of the codebase IMO.

Thanks! that was a much needed cleaning :) Removing them like you did works for me.

> As for the documentation, `ProjectionLike` documentation is not generated with JSDoc but is generated with TypeDoc. I however did change our documentation engine for now, since it will introduce a lot of conflicts with #2414. Waiting for it to be merged!

Ok, thanks.

Desplandis commented:
Added some breaking change notes in the commit messages.
